File: api/src/integration/java/org/pmiops/workbench/google/CloudResourceManagerServiceImplIntegrationTest.java
package org.pmiops.workbench.google;

import static com.google.common.truth.Truth.assertThat;

import com.google.api.services.cloudresourcemanager.model.Project;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.BaseIntegrationTest;
import org.pmiops.workbench.db.model.DbUser;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Import;

public class CloudResourceManagerServiceImplIntegrationTest extends BaseIntegrationTest {
  @Autowired private CloudResourceManagerService service;

  @TestConfiguration
  @ComponentScan(basePackageClasses = CloudResourceManagerServiceImpl.class)
  @Import({CloudResourceManagerServiceImpl.class, BaseIntegrationTest.Configuration.class})
  static class Configuration {}

  // This is a single hand created user in the fake-research-aou.org gsuite.
  // It has one project that has been shared with it, AoU CRM Integration Test
  // in the firecloud dev domain.
  private final String SINGLE_WORKSPACE_TEST_USER_EMAIL =
      "cloud-resource-manager-integration-test@fake-research-aou.org";
  // This user was hand-created in the test environment and has never signed in.
  private final String NEVER_SIGNED_IN_TEST_USER_EMAIL =
      "cloud-resource-manager-integration-test-no-sign-in@fake-research-aou.org";

  @Test
  public void testGetAllProjectsForUserSingleWorkspaceUser() throws Exception {
    DbUser testUser = new DbUser();
    testUser.setUsername(SINGLE_WORKSPACE_TEST_USER_EMAIL);
    List<Project> projectList = service.getAllProjectsForUser(testUser);
    assertThat(projectList.size()).isEqualTo(1);
  }

  @Test
  public void testGetAllProjectsForUserNeverSignedInUser() throws Exception {
    DbUser testUser = new DbUser();
    testUser.setUsername(NEVER_SIGNED_IN_TEST_USER_EMAIL);
    List<Project> projectList = service.getAllProjectsForUser(testUser);
    assertThat(projectList).isEmpty();
  }
}

File: api/src/integration/java/org/pmiops/workbench/google/CloudResourceManagerServiceImplIntegrationTest.kt
package org.pmiops.workbench.google

import com.google.common.truth.Truth.assertThat

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
import com.google.api.client.http.apache.ApacheHttpTransport
import com.google.api.services.cloudresourcemanager.model.Project
import java.io.IOException
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.pmiops.workbench.IntegrationTestConfig
import org.pmiops.workbench.auth.Constants
import org.pmiops.workbench.auth.ServiceAccounts
import org.pmiops.workbench.db.model.User
import org.pmiops.workbench.test.Providers
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.retry.backoff.NoBackOffPolicy
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@ContextConfiguration(classes = [IntegrationTestConfig::class])
class CloudResourceManagerServiceImplIntegrationTest {
    private var service: CloudResourceManagerServiceImpl? = null
    private val httpTransport = ApacheHttpTransport()

    // This is a single hand created user in the fake-research-aou.org gsuite.
    // It has one project that has been shared with it, AoU CRM Integration Test
    // in the firecloud dev domain.
    private val CLOUD_RESOURCE_MANAGER_TEST_USER_EMAIL = "cloud-resource-manager-integration-test@fake-research-aou.org"

    // N.B. this will load the default service account credentials for whatever AoU environment
    // is set when running integration tests. This should be the test environment.
    @Autowired
    @Qualifier(Constants.DEFAULT_SERVICE_ACCOUNT_CREDS)
    private var serviceAccountCredential: GoogleCredential? = null

    @Autowired
    private val serviceAccounts: ServiceAccounts? = null

    @Autowired
    @Qualifier(Constants.CLOUD_RESOURCE_MANAGER_ADMIN_CREDS)
    private val cloudResourceManagerAdminCredential: GoogleCredential? = null

    @Before
    @Throws(IOException::class)
    fun setUp() {
        // Get a refreshed access token for the CloudResourceManager service account credentials.
        serviceAccountCredential = serviceAccountCredential!!.createScoped(CloudResourceManagerServiceImpl.SCOPES)
        serviceAccountCredential!!.refreshToken()
        service = CloudResourceManagerServiceImpl(
                Providers.of(cloudResourceManagerAdminCredential),
                httpTransport,
                GoogleRetryHandler(NoBackOffPolicy()),
                serviceAccounts)
    }

    @Test
    fun testGetAllProjectsForUser() {
        val testUser = User()
        testUser.email = CLOUD_RESOURCE_MANAGER_TEST_USER_EMAIL
        val projectList = service!!.getAllProjectsForUser(testUser)
        assertThat(projectList.size).isEqualTo(1)
    }
}
